Here is the page that has list of all available bert models on tensorflow hub that one can download and make use of.

https://tfhub.dev/google/collections/bert/1

Here is the information on basic uncased BERT model,

https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4

It uses L=12 hidden layers (i.e., Transformer blocks), a hidden size of H=768, and A=12 attention heads. This model has been pre-trained for English on the Wikipedia and BooksCorpus.

In [1]:
!pip install tensorflow_text

In [2]:

import tensorflow_hub as hub
import tensorflow_text as text

In [3]:
preprocess_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3"
encoder_url = "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4"

In [4]:
bert_preprocess_model = hub.KerasLayer(preprocess_url)

In [17]:
text_test = ['nice movie indeed . i have to mucch work ','I love python programming']
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_type_ids', 'input_mask', 'input_word_ids'])

In [18]:
text_preprocessed['input_mask']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [19]:
text_preprocessed['input_type_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [20]:
text_preprocessed['input_word_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[  101,  3835,  3185,  5262,  1012,  1045,  2031,  2000, 14163,
         9468,  2232,  2147,   102,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0, 

**101 --> CLS token**

**102 --> SEP token**

BERT uses CLS as a special token at the begining of each setence whereas SEP as a special token to
separate two sentences or end sinle sentece

In [21]:
bert_model = hub.KerasLayer(encoder_url)

In [22]:
bert_results = bert_model(text_preprocessed)

In [23]:
bert_results.keys()

dict_keys(['pooled_output', 'sequence_output', 'encoder_outputs', 'default'])

In [24]:
bert_results['sequence_output']

# it will represent 2 sentences,  128 --> token size, each token convert into 768--> embeddings'

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[ 0.04481602,  0.15432283,  0.24396122, ..., -0.09724917,
          0.2678497 ,  0.47033936],
        [ 0.27531156, -0.2107404 ,  1.1524328 , ..., -0.01355545,
          0.32640818, -0.00918693],
        [ 0.29774082, -0.31329605, -0.07567675, ..., -0.01907246,
          0.01683251, -0.44119906],
        ...,
        [-0.22420658, -0.19053377,  0.6908112 , ...,  0.60566366,
          0.07286917,  0.06078712],
        [-0.21383154, -0.32266745,  0.5345733 , ...,  0.7357798 ,
          0.18004218,  0.01525757],
        [-0.03208486, -0.02116879,  0.77698743, ...,  0.40653718,
          0.02995573,  0.01373504]],

       [[-0.0790057 ,  0.3633508 , -0.21101609, ..., -0.17183742,
          0.16299754,  0.6724269 ],
        [ 0.27883506,  0.43716297, -0.35764715, ..., -0.04463629,
          0.38315165,  0.5887988 ],
        [ 1.2037677 ,  1.0727019 ,  0.4840876 , ...,  0.24920993,
          0.4073093 ,  0.404818  ],
        ...,

In [25]:
bert_results['pooled_output']

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.8958751 , -0.36458364, -0.6555456 , ..., -0.41266897,
        -0.63063115,  0.93713784],
       [-0.917123  , -0.4793516 , -0.7865695 , ..., -0.6175171 ,
        -0.71026856,  0.9218428 ]], dtype=float32)>

In [26]:
len(bert_results['encoder_outputs'])

12

Since we are using BERT base model it has 12 encoder layers, that's why above the length of encoder_output is 12.
Read more about purpose of each element here: https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4

You can see below that last element of encoder_outputs is basically a sequence_output

In [27]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']


# The sequence_output is same as output of 12th layer of encoder_outputs

<tf.Tensor: shape=(2, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]]])>

In [29]:
bert_results['encoder_outputs']    # 12 LAYERS

[<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
 array([[[ 0.00427691,  0.06076157, -0.17014372, ...,  0.22486575,
          -0.09739947, -0.0764861 ],
         [ 1.2259893 ,  1.217819  ,  1.074299  , ...,  0.1351935 ,
          -0.27399352, -0.36679125],
         [ 0.00693834,  0.6447532 , -0.4278963 , ..., -0.08671824,
           0.84340805, -1.1842225 ],
         ...,
         [-0.07261764, -0.25106007,  0.6435405 , ...,  0.42317277,
          -0.25080454,  0.06628355],
         [-0.16749042, -0.26670915,  0.52496356, ...,  0.4440205 ,
          -0.17575118, -0.03052453],
         [-0.03202928, -0.18227012,  0.54503   , ...,  0.74391866,
          -0.46544263, -0.1000725 ]],
 
        [[ 0.18903571,  0.02752544, -0.06513733, ..., -0.00620206,
           0.1505388 ,  0.03165448],
         [ 0.5916147 ,  0.758914  , -0.07240676, ...,  0.6190399 ,
           0.8292891 ,  0.16161981],
         [ 1.4460828 ,  0.44602662,  0.40990242, ...,  0.48255903,
           0.62691176,  0.13